scalyr_agent/util.py

```diff
@@ -1337,3 +1341,265 @@ def close(self):
+        self._cluster_rate_lock = threading.Lock()
+
+        self._max_concurrency = max_concurrency
+
+        # A queue of tokens
+        self._ripe_time = None
+        self._token_queue = deque()
+        # Condition variable to synchronize access to token queue
+        self._token_queue_cv = threading.Condition()
+
+        # fake clock for testing
+        self._fake_clock = fake_clock
+        self._test_mode_lock = threading.Lock()
+
+        # Validate input
+        strategies = [self.INCREASE_STRATEGY_MULTIPLY, self.INCREASE_STRATEGY_RESET_THEN_MULTIPLY]
+        if increase_strategy not in strategies:
+            raise ValueError('Increase strategy must be one of %s' % strategies)
+
+        if self._max_concurrency < 1:
+            raise ValueError('max_concurrency must be greater than 1')
+
+        if self._consecutive_success_threshold <= 1:
+            raise ValueError(
+                'consecutive_success_threshold must be a positive int. Value=%s' % consecutive_success_threshold)
+
+        if self._initial_cluster_rate < self._lower_cluster_rate or self._initial_cluster_rate > self._upper_cluster_rate:
+            raise ValueError(
+                'Initial cluster rate must be between lower and upper rates. Initial=%s, Lower=%s, Upper=%s.' % (
+                    initial_cluster_rate, lower_cluster_rate, upper_cluster_rate))
+
+        self._initialize_token_queue()
+
+    @property
+    def current_cluster_rate(self):
+        return self._current_cluster_rate
+
+    @property
+    def current_agent_rate(self):
+        return float(self._current_cluster_rate) / self._num_agents
+
+    def _initialize_token_queue(self):
+        """Initialize a queue with max_concurrency tokens."""
+        self._token_queue_cv.acquire()
+        for num in range(self._max_concurrency):
+            token = RateLimiterToken(num)
+            self._token_queue.append(token)
+        self._ripe_time = self._get_next_ripe_time()
+        self._token_queue_cv.release()
+
+    def _adjust_cluster_rate(self, success):
+        """Adjust cluster rate, backing off on failure or increasing on enough consecutive successes.
+
+        Any failure will cause backoff.
+        A success will accumulate until self._consecutive_success_threshold is reached, at which point the rate will be
+            increased according to self._increase_strategy.  When this happens, the success accumulator resets to 0
+
+        @param success: Whether an operation was successful
+        @type success: bool
+        """
+        self._cluster_rate_lock.acquire()
+        try:
+            if success:
+                self._consecutive_successes += 1
+            else:
+                self._consecutive_successes = 0
+            if not success:
+                # Failure case: halve the rate
+                self._current_cluster_rate = max(
+                    self._lower_cluster_rate,
+                    self._current_cluster_rate / self._backoff_factor
```

Comment:
Most times I see a backoff factor in these type of algorithms, it is expressed as a multiplication instead of a division.  so, instead of having a backoff factor of 2, have it be 0.5.  Can you make that change?